File: src/main/java/org/hunter/courseschedule.java
package org.hunter;

import java.util.*;

class courseschedule {

    public boolean canFinish(int numCourses, int[][] prerequisites) {
        Map<Integer,List<Integer>> graph = new HashMap<>();
        for(int [] preq : prerequisites){
            graph.putIfAbsent(preq[0], new ArrayList<>());
            graph.get(preq[0]).add(preq[1]);
        }
        Set<Integer> visited = new HashSet<>();
        for(int i = 0; i < numCourses; ++i){
            if(!dfs(graph, i, new HashSet<>(), visited)){
                return false;
            }
        }
        return true;
    }

    boolean dfs(Map<Integer,List<Integer>> graph, int course, Set<Integer> visited, Set<Integer> globalVisited){
        if(visited.contains(course)){
            return false;
        }
        if(globalVisited.contains(course)){
            return true;
        }        
        globalVisited.add(course);
        visited.add(course);
        for(int p : graph.getOrDefault(course, List.of())){
            if(!dfs(graph, p, visited, globalVisited)){
                return false;
            }
        }
        visited.remove(course);
        return true;
    }
}

File: course_schedule.py
#[[1,0],[0,1]]
class Solution:
    def canFinish(self, numCourses: int, prerequisites: list[list[int]]) -> bool:
        adj = {}
        seen = set()
        for pre in prerequisites:
            course = pre[0]
            p = pre[1]
            if course not in adj:
                adj[course] = []
            adj[course].append(p)

        for course in range(numCourses):
            if course not in adj:
                adj[course] = []

        for course in range(numCourses):
            if not self.dfs(adj, course, set(), seen):
                return False

        return True

    def dfs(self, adj, course, path, seen):
        if course in path:
            return False

        if course in seen:
            return True

        path.add(course)
        seen.add(course)
        ans = True

        for pre in adj[course]:
            if not self.dfs(adj, pre, path, seen):
                return False

        path.remove(course)
        return ans